### PR TITLE
feat: native sandbox mode for process-level isolation

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

Adds kernel-enforced process isolation without containers/VMs. Implements native sandbox mode for improved performance and simplified deployment.

- Linux: User namespaces (unshare + pivot_root) — no root needed
- macOS: Dedicated system user (_mino_agent) + pf + privileged helper binary
- Filtering SOCKS5/HTTP CONNECT proxy for --network-allow
- Config-driven: `runtime = "native"` in TOML, `--runtime native` CLI flag
- Full CLI integration: run, stop, list, logs, exec, status, setup --native

## Changes

- **Core**: User namespace isolation (Linux), privileged helper (macOS)
- **Network**: SOCKS5/HTTP CONNECT proxy filtering for --network-allow
- **CLI**: Runtime flag, setup command updates
- **Tests**: 221 new tests (731 total), full coverage of native runtime

## Breaking Changes

None

## Testing

- 221 new tests added
- 731 total tests passing
- clippy clean
- Integration tests cover native runtime operations (run, exec, stop, logs)
- Manual testing on Linux and macOS

## Related Issues

Part of native sandbox implementation roadmap.

### CODE REVIEW

File: src/bin/mino-sandbox-helper.rs [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+}
+
+/// Set up signal forwarding to child process
+///
+/// # Safety
+/// Installs C-style signal handlers via libc::signal.
+/// Must be called only once, from the parent process after fork().
+#[cfg(unix)]
+unsafe fn setup_signal_forwarding(child_pid: i32) {
+    CHILD_PID.store(child_pid, Ordering::SeqCst);
```
Comment (dean0x) [inline src/bin/mino-sandbox-helper.rs]: ⚠️ **HIGH**: Unsafe libc::signal cast is non-portable (85% confidence)

The cast through `*const () -> usize` may fail on platforms where `libc::sighandler_t` is not `usize`. Also, `signal()` behavior varies across platforms regarding handler reset.

**Risk**: UB or compilation failures on non-macOS systems.

**Fix**: Use `libc::sigaction` with `SA_RESTART` for POSIX compliance:
```rust
unsafe fn setup_signal_forwarding(child_pid: i32) {
    CHILD_PID.store(child_pid, Ordering::SeqCst);
    let mut sa: libc::sigaction = std::mem::zeroed();
    sa.sa_sigaction = forward_signal as usize;
    sa.sa_flags = libc::SA_RESTART;
    libc::sigaction(libc::SIGINT, &sa, std::ptr::null_mut());
    libc::sigaction(libc::SIGTERM, &sa, std::ptr::null_mut());
}
```

File: src/sandbox/macos.rs [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+        .await;
+
+    match result {
+        Ok(status) if status.success() => Ok(()),
+        _ => Err(MinoError::SandboxNotSetup),
+    }
+}
+
+async fn check_helper_version() -> MinoResult<()> {
+    let output = Command::new(HELPER_BINARY)
```
Comment (dean0x) [inline src/sandbox/macos.rs]: ⚠️ **HIGH**: cleanup() hardcodes sandbox_user (92% confidence)

`MacosSandbox::cleanup()` calls `SandboxConfig::default().sandbox_user` instead of using the actual configured user. If user sets `sandbox_user = "custom-agent"`, ACLs are set for that user during spawn but cleanup tries to remove for `_mino_agent`, leaking stale ACLs.

**Risk**: ACL leaks for projects using custom sandbox users.

**Fix**: Extend trait signature to accept sandbox_user:
```rust
async fn cleanup(&self, session_id: &str, project_dir: &Path, sandbox_user: &str) -> MinoResult<()>;
```

File: src/bin/mino-sandbox-helper.rs [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+        return Err(format!("ACL setup failed on home dir: {}", e));
+    }
+
+    // 4. Look up sandbox user UID and GID
+    let ready = prepare_spawn(&sandbox_user, &acl_paths, &home_dir)?;
+
+    // 5. Fork + setgid + setuid + exec
+    #[cfg(unix)]
+    {
+        // SAFETY: fork() is the only FFI call here
```
Comment (dean0x) [inline src/bin/mino-sandbox-helper.rs]: **[RUST - HIGH, 85% confidence]** SAFETY comment on `fork()` is inaccurate. Current comment: "fork() is the only FFI call here" is a tautology. The actual safety requirement is that the helper is single-threaded (no locked mutexes).

Fix:
```rust
// SAFETY: fork() duplicates the process. The helper binary is single-threaded
// (no tokio runtime, no background threads), so there is no risk of duplicating
// locked mutexes. The child calls only async-signal-safe libc functions.
let pid = unsafe { libc::fork() };
```

**Comment by dean0x** [comment]
## Code Review Summary - PR #58: Native Sandbox Support

Comprehensive review across 8 reviewers (architecture, complexity, consistency, performance, regression, Rust, security, tests).

### High-Confidence Findings by Severity

**Total Findings**: 36 high-confidence items (≥80%), plus 15 medium-confidence suggestions

#### CRITICAL: 0
#### HIGH: 23 items
- **Security (2)**: TOCTOU race in write_restricted_file, case-insensitive hostname bypass
- **Performance (4)**: HTTP CONNECT heap allocation, linear scan in is_allowed(), sequential stale cleanup, subprocess call overhead
- **Architecture (3)**: Duplicate username validation (3 copies), setup.rs exceeds 1400 lines
- **Consistency (3)**: Result<T, String> vs MinoResult<T>, CacheMode serde inconsistency, #[cfg] vs trait dispatch
- **Complexity (5)**: generate_setup_script 134 lines, check_podman_in_vm 102 lines, check_native_podman 96 lines, check_rootless_mode_in_vm 101 lines, drop_privileges duplication
- **Rust (3)**: Missing SAFETY comment on getuid(), inaccurate SAFETY comment on fork(), silent copy failures
- **Tests (2)**: Flaky sleep-based test, signal tests don't verify signal type behavior

#### MEDIUM: 13 items
- **Regression (1)**: resolve_runtime_mode early rejection of unrecognized config values (82% confidence)
- **Architecture (2)**: SandboxPlatform/ContainerRuntime parallel traits (82%), SandboxSpawnConfig parameter bundling (80%)
- **Complexity (4)**: uninstall_native_macos repetitive cleanup, setup_sandbox_user 82 lines, configure_sudoers 77 lines, setup.rs module split needed
- **Performance (2)**: Sequential setup checks on macOS, cleanup_macos_sandbox temp file I/O
- **Rust (1)**: proxy_env_vars clone overhead (83%)
- **Security (3)**: Request file cleanup, path canonicalization, defensive validation in get_user_ids (80%)
- **Tests (3)**: prepare_dotfiles untested, wait_with_signal_forwarding untested, weak platform assertion on Linux (85%)

---

### Medium-Confidence Suggestions (60-79%)

Collected in summary (not posted as inline comments per specification):

1. **Proxy module organization** (70%) - 1167-line file could split into submodules
2. **Helper binary organization** (65%) - 808-line helper mixes concerns (arg parsing, IPC, fork/exec, signals)
3. **ContainerRuntime trait size** (70%) - 20+ methods could split into lifecycle/image/volume managers
4. **Distro package manager duplication** (75%) - Logic repeated across 3 setup functions
5. **Visibility inconsistency** (Medium flagged above)
6. **Setup.rs validation generation** (70%) - _session_id unused parameter in generate_pf_rules
7. **No idle timeout on relay tunnels** (68%) - Stalled connections hold semaphore indefinitely
8. **mino list table layout** (65%) - RUNTIME column shifts positional parsing (backward compat issue)
9. **New [sandbox] config section** (60%) - Always appears in serialized output even when unused
10. **Commands enum lint suppression** (70%) - #[allow(clippy::large_enum_variant)] on Run variant; consider boxing instead
11. **remove_acl subprocess redundancy** (62%) - Always tries both rw and ro removal regardless of original type
12. **generate_setup_script format intermediates** (70%) - format!() allocations in tight loop
13. **Session backward compatibility** (Verified - 98% confidence) - All new fields use Option<T> + #[serde(default)]
14. **Detect package manager logic** (75%) - run_visible_sudo/run_visible_orb return value handling could be consolidated
15. **accept_loop nesting** (62%) - 5 levels of nesting in async accept loop (standard pattern, difficult to simplify)

---

### Recommendations by Category

**SECURITY** - Address both HIGH items immediately:
1. Fix TOCTOU in write_restricted_file (check for symlink before removing)
2. Normalize hostname to lowercase in is_allowed()

**PERFORMANCE** - Fix both HIGH items:
1. Switch HTTP CONNECT from heap to stack buffer (single-line fix)
2. Convert is_allowed() rule lookup to HashMap<String, HashSet<u16>> (O(1) vs O(N))

**ARCHITECTURE/CONSISTENCY** - Consolidate validation:
- Remove validate_pf_username(), call validate_sandbox_user() from generate_pf_rules()
- Reuse validate_sandbox_user() in configure_sudoers instead of inline check

**COMPLEXITY** - Can be addressed in follow-up PR:
- Extract append_bind_mount() in generate_setup_script (eliminates passthrough/writable duplication)
- Extract drop_privileges() in helper binary (security-critical code, flagged 3 consecutive reviews)
- Plan setup.rs module split (1499 lines, 5 distinct domains)

**TESTS** - Address both HIGH items:
1. Replace sleep-based flaky test with polling retry loop
2. Rename or strengthen signal tests to verify actual signal behavior

**REGRESSION** - Low priority (field is new, defaults correctly):
- Consider lenient fallback for unrecognized runtime values in config (warn instead of error)

---

### Review Statistics

| Review | CRITICAL | HIGH | MEDIUM | LOW | Score |
|--------|----------|------|--------|-----|-------|
| Architecture | 0 | 2 | 2 | 0 | 8/10 |
| Complexity | 0 | 4 | 4 | 0 | 7/10 |
| Consistency | 0 | 2 | 1 | 0 | 7/10 |
| Performance | 0 | 2 | 2 | 0 | 7/10 |
| Regression | 0 | 0 | 1 | 0 | 9/10 |
| Rust | 0 | 2 | 2 | 0 | 8/10 |
| Security | 0 | 2 | 2 | 0 | 8/10 |
| Tests | 0 | 2 | 2 | 1 | 8/10 |
| **TOTAL** | **0** | **23** | **13** | **0** | **7.6/10** |

---

### Inline Comments Posted

16 inline comments created on PR #58:
- src/sandbox/macos.rs (4): TOCTOU race, case-insensitive hostname, request cleanup, unnecessary clones
- src/sandbox/proxy.rs (3): HTTP CONNECT allocation, linear scan, string clones
- src/sandbox/linux.rs (2): Missing SAFETY comment, uncanonicalized paths
- src/sandbox/helper.rs (2): Result vs MinoResult, silent copy failures
- src/sandbox/native.rs (1): Weak platform assertion
- src/sandbox/config.rs (1): CacheMode serde inconsistency
- src/bin/mino-sandbox-helper.rs (1): Inaccurate SAFETY comment, drop_privileges duplication
- src/cli/commands/run/native.rs (3): HOME inconsistency, prepare_dotfiles untested, wait_with_signal_forwarding untested
- src/cli/commands/run/mod.rs (1): Early resolve_runtime_mode validation
- src/cli/commands/setup.rs (2): Inline validation duplication, complexity issues
- src/cli/commands/stop.rs (1): Signal test verification
- src/cli/commands/status.rs (1): #[cfg] vs trait dispatch

---

### Positive Strengths

**Well-executed patterns:**
- Proper MinoResult<T> / MinoError with .hint() for 7 new error variants
- Clean SandboxPlatform trait with compile-time dispatch via #[cfg]
- 11 unsafe blocks, mostly with SAFETY comments
- AtomicI32 for signal handler state (no static mut)
- Shell injection prevention via shell_quote + escape_path
- RAII Drop pattern for ProxyHandle
- 731 tests covering protocols, injection safety, boundary values
- Protocol-level integration tests for proxy and sandbox behavior

---

### Next Steps

**Required before merge:**
1. Fix TOCTOU race in write_restricted_file (SECURITY)
2. Normalize hostname comparison (SECURITY)
3. Fix HTTP CONNECT buffer allocation (PERFORMANCE)
4. Convert is_allowed() rule lookup (PERFORMANCE)

**Strongly recommended for this PR:**
5. Add SAFETY comment on getuid() (RUST)
6. Improve SAFETY comment on fork() (RUST)
7. Consolidate username validation (ARCHITECTURE)

**Can defer to follow-up PR:**
8. Extract helpers in setup.rs, generate_setup_script, helper binary (COMPLEXITY/MAINTENANCE)
9. Strengthen test coverage for prepare_dotfiles, signal handling (TESTS)
10. Address MEDIUM-confidence items (will not block merge if issues are minor)

---

**Claude Code - Automated Code Review**
*Review based on 8 independent reviewer passes across architecture, complexity, consistency, performance, regression, Rust patterns, security, and test coverage.*

File: src/cli/commands/run/native.rs [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+    env.extend(config.container.env.clone());
+
+    env
+}
+
+/// Resolve project directory from CLI args or current directory
+fn resolve_project_dir(args: &RunArgs) -> MinoResult<PathBuf> {
+    let dir = match &args.project {
+        Some(p) => p.clone(),
+        None => {
```
Comment (dean0x) [inline src/cli/commands/run/native.rs]: 🔨 **Architecture: Duplicated  function**

`resolve_project_dir` is implemented twice with divergent logic:
- Container version (mod.rs:625): canonicalizes the path for symlink resolution
- Native version (native.rs:494): simple `is_dir()` check without canonicalization

This DRY violation means bugs in one won't be caught in the other, and symlinked project directories behave differently between modes.

**Confidence:** 95%

**Fix:** Extract a shared `resolve_project_dir` in the parent `run/mod.rs` module and call it from both paths. The container version's canonicalization is the correct behavior to keep.
